src/raftstore/store/snap/builder.rs

```diff
@@ -0,0 +1,465 @@
+                    cf_size += l;
+                    if let Some(ref limiter) = io_limiter {
+                        if bytes >= base {
+                            bytes = 0;
+                            limiter.request(base);
+                        }
+                        bytes += l as i64;
+                    }
+                    writer.put(key, value)?;
+                    Ok(!stale_for_generate(snap_key, snap_stale_notifier))
+                })?;
+                if cf_key_count > 0 {
+                    writer.finish()?;
+                }
+            }
+        }
+
+        // TODO: calculate written bytes and crc32 while scaning.
+        self.written_bytes = get_file_size(&self.tmp_cf_path)?;
+        self.digest = Digest::new_with_initial(crc32::IEEE, calc_crc32(&self.tmp_cf_path)?);
+
+        SNAPSHOT_CF_KV_COUNT
+            .with_label_values(&[self.cf])
+            .observe(cf_key_count as f64);
+        SNAPSHOT_CF_SIZE
+            .with_label_values(&[self.cf])
+            .observe(cf_size as f64);
+        info!(
+            "[region {}] scan snapshot on cf {}, key count {}, size {}",
+            region_id, self.cf, cf_key_count, cf_size
+        );
+        Ok(cf_key_count)
+    }
+
+    fn save(&self, to: &PathBuf) -> io::Result<()> {
+        if self.written_bytes > 0 {
+            fs::rename(&self.tmp_cf_path, &to)?;
+        } else {
+            delete_file_if_exist(&self.tmp_cf_path);
+        }
+        Ok(())
+    }
+}
+
+struct Inner {
+    dir: String,
+    for_send: bool,
+    key: SnapKey,
+
+    io_limiter: Option<Arc<IOLimiter>>,
+    tmp_meta_path: PathBuf,
+    tmp_meta_file: File,
+    tmp_cf_files: Vec<CfFile>,
+    hold_tmp_files: bool,
+}
+
+impl Inner {
+    fn new(
+        dir: String,
+        for_send: bool,
+        key: SnapKey,
+        io_limiter: Option<Arc<IOLimiter>>,
+    ) -> Result<Inner> {
+        let mut open_options = OpenOptions::new();
+        open_options.write(true).create_new(true);
+        let tmp_meta_path = get_meta_tmp_file_path(&dir, for_send, key);
+        let tmp_meta_file = open_options.open(&tmp_meta_path)?;
+        Ok(Inner {
+            dir,
+            for_send,
+            key,
+            io_limiter,
+            tmp_meta_path,
+            tmp_meta_file,
+            tmp_cf_files: Vec::new(),
+            hold_tmp_files: true,
+        })
+    }
+
+    fn save(&mut self) -> Result<SnapshotMeta> {
+        let mut snapshot_meta = SnapshotMeta::new();
+        for cf_builder in &self.tmp_cf_files {
+            let to = get_cf_file_path(&self.dir, self.for_send, self.key, cf_builder.cf);
+            cf_builder.save(&to)?;
+
+            let mut snap_cf_file = SnapshotCFFile::new();
+            snap_cf_file.set_cf(cf_builder.cf.to_owned());
+            snap_cf_file.set_size(cf_builder.written_bytes);
+            snap_cf_file.set_checksum(cf_builder.digest.sum32());
+            snapshot_meta.mut_cf_files().push(snap_cf_file);
+        }
+
```

Comment:
Please add more comments for each step to make it more readable.

Reply:
fixed.